SetColumnName/SetRowName to support extending document adding new columns/rows

Sven-HP opened the pull request:
SetColumnName: allow setting column names for an empty table / empty column

Hi @d99kris 
Thank you for rapidcsv! I love the easy-to-use interface.

My use-case creates a table (no CSV import), then exports to CSV. I need to prepare the table and set the column names _before_ adding any data. This PR avoids the crash.
Tested on Windows with Visual Studio 2017.

Note: I did not change the version number in line 5.

Note2: Maybe need a similar fix for SetRowName?

d99kris commented:
Hi @Sven-HP - thanks for the positive feedback, and for the proposed contribution!

Generally speaking I think rapidcsv needs _something_ like this or similar. I'm just thinking whether it makes more sense to extend existing Set* functions, or whether a new API function should be added. One potential use-case could be constructing a CSV document from scratch **without** column headers. Then we'd need to extend SetColumn itself (and for consistency SetRow, SetCell). 

To avoid that I'm thinking it could be simpler to introduce a new function that allows setting the CSV document size. Perhaps `rapidcsv::Resize(size_t pColumns, size_t pRows)` that would allow setting number of data columns/rows, and then rapidcsv would set up the corresponding data structure. pRows could maybe be an optional parameter with default value 0 (or 1).

Would having such an API function solve the problem you are facing? If yes, I can try prepare a patch.

Sven-HP commented:
Hi @d99kris 
Thank you. I also like your `rapidcsv::Resize(size_t pColumns, size_t pRows)` suggestion. And yes, it would also solve my problem.

However I slightly prefer my suggestion because:
- I think it is more consistent with the existing code: If am not mistaken, the other Set* functions SetRow, SetCell, SetColumn are already robust against "too large" idx, meaning they automatically increase the table size(s) when the given idx parameters require it. 
- It is more robust against crashes when a "new" / "too large" index is used "accidentally".

For the use case you mention, "constructing a CSV document from scratch without column headers",
- no special action is necessary, because SetRow/SetColumn/SetCell function already accept and store the data as they come.
- your suggested Resize() would not hurt either, might be more efficient for very large tables.

Please let me know what you think. Of course it's your decision, both ways are fine. :-)

Thanks
Sven

d99kris commented:
Hi @Sven-HP 
Thanks a lot for refreshing my memory. You are completely right, all the other Set* functions handle "too large" idx, and as such it makes perfect sense for SetColumnName() to support the same. 

I will add some minor suggestions in the PR - mainly to be consistent with the other code in rapidcsv.h. In addition to that I think it would be great if we could cover the functionality with a test. Perhaps a new test tests/test074.cpp based on tests/test021.cpp could be added (just change order and do SetColumnName before setting columns for example). If you're not comfortable adding the test I can add it myself after merging the PR.

d99kris commented on src/rapidcsv.h:
```diff
@@ -1068,6 +1068,12 @@ namespace rapidcsv
       }
 
+      // increase table size if necessary:
+      int rowIdx = mLabelParams.mColumnNameIdx;
```
Propose to make this: const int rowIdx ....

d99kris commented on src/rapidcsv.h:
```diff
@@ -1068,6 +1068,12 @@ namespace rapidcsv
 
+      // increase table size if necessary:
+      int rowIdx = mLabelParams.mColumnNameIdx;
+      if ((int)mData.size() <= rowIdx) mData.resize(rowIdx + 1);
```
Current code is not wrong, but can we change the order (so it matches other rapidcsv code), i.e.
```if (rowIdx > (int)mData.size())```

d99kris replied:
Also can we change the code formatting to something like:
```
if (rowIdx > (int)mData.size())
{
  mData.resize(rowIdx + 1);
}
```

d99kris commented on src/rapidcsv.h:
```diff
@@ -1068,6 +1068,12 @@ namespace rapidcsv
+      int rowIdx = mLabelParams.mColumnNameIdx;
+      if ((int)mData.size() <= rowIdx) mData.resize(rowIdx + 1);
+      auto& row = mData[rowIdx];
+      if ((int)row.size() <= columnIdx) row.resize(columnIdx + 1);
```
Same as above comment, propose to change into something like:
```
if (columnIdx > (int)row.size())
{
  row.resize(columnIdx + 1);
}
```

Sven-HP commented:
Hi @d99kris Sorry I should have followed your code style. I have adjusted the style now.

Also I have added the same checks / table size increases to SetRowName.

Would be great if you could write the tests - sorry I have no experience with ctest (would be a great opportunity to learn, but I currently don't have enough time...)

I suggest to add the tests _before_ merging into master.

d99kris commented:
> Also I have added the same checks / table size increases to SetRowName.

Great, thanks! 

> Would be great if you could write the tests - sorry I have no experience with ctest (would be a great opportunity to learn, but I currently don't have enough time...)

Sure, no worries!

> I suggest to add the tests before merging into master.

Yeah that would generally be preferable, but I think it's fine. I suspect you might be the only real user currently using the library to construct a CSV file. I will soon add the test to master, and only after that I'll bump the version and tag a release. I expect users to download the release packages if getting a stable version is important. :)

Sven-HP commented:
Perfect - thank you! 👍
